suite/tests/client-interface/translate_sandbox.c

```diff
@@ -0,0 +1,145 @@
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef ASM_CODE_ONLY /* C code */
+#include "tools.h" /* for print() */
+
+#include <assert.h>
+
+int sandbox();
+int usebx();
+
+/* top-level exception handler */
+static LONG
+our_top_handler(struct _EXCEPTION_POINTERS * pExceptionInfo)
+{
+    if (pExceptionInfo->ExceptionRecord->ExceptionCode == EXCEPTION_ACCESS_VIOLATION) {
+        print("access violation exception\n");
+        protect_mem(usebx, 1024, ALLOW_READ|ALLOW_WRITE|ALLOW_EXEC);
+        return EXCEPTION_CONTINUE_EXECUTION;
+    }
+    return EXCEPTION_EXECUTE_HANDLER; /* => global unwind and silent death */
+}
+
+int main(int argc, char *argv[])
+{
+    INIT();
+    int count = 0;
+
+    SetUnhandledExceptionFilter((LPTOP_LEVEL_EXCEPTION_FILTER) our_top_handler);
+
+    print("start of test, count = %d\n", count);
+    protect_mem(sandbox, 1024, ALLOW_READ|ALLOW_WRITE|ALLOW_EXEC);
+    protect_mem(usebx, 1024, ALLOW_READ);
+    count = sandbox();
+    count += usebx();
+    print("end of test, count = %d\n", count);
+    return 0;
+}
+
+
+#else /* asm code *************************************************************/
+#include "asm_defines.asm"
+START_FILE
+
+#if defined(ASSEMBLE_WITH_GAS)
+# define ALIGN_WITH_NOPS(bytes) .align (bytes)
+# define FILL_WITH_NOPS(bytes)  .fill  (bytes), 1, 0x90
+#elif defined(ASSEMBLE_WITH_MASM)
+/* Cannot align to a value greater than the section alignment.  See .mytext
+ * below.
+ */
+# define ALIGN_WITH_NOPS(bytes) ALIGN bytes
+# define FILL_WITH_NOPS(bytes) \
+    REPEAT (bytes) @N@\
+        nop @N@\
+        ENDM
+#else
+# define ALIGN_WITH_NOPS(bytes) ALIGN bytes
+# define FILL_WITH_NOPS(bytes) TIMES bytes nop
+#endif
+
+#if defined(ASSEMBLE_WITH_MASM)
+/* The .text section on Windows is only 16 byte aligned.  If we use plain
+ * ALIGN_WITH_NOPS(), we get error A2189.  We can't re-declare .text, but we can
+ * declare a new section with greater alignment.  The EXECUTE gets the right
+ * protections and makes ALIGN_WITH_NOPS() do nop fill.
+ */
+_MYTEXT SEGMENT ALIGN(4096) READ EXECUTE SHARED ALIAS(".mytext")
+#endif
+
+/* int sandbox()
+ *   Modifies self to enter sandbox.
+ *   Then does another self modification in another page.
+ *   Should return 2 (using ebx).
+ */
+
+#define FUNCNAME sandbox
+DECLARE_FUNC(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+        mov      REG_XAX, HEX(1)
+        lea      REG_XDX, SYMREF(sandbox_immediate_addr_plus_four - 4)
+        mov      DWORD [REG_XDX], eax        /* selfmod write */
+        mov      REG_XDX, HEX(0)             /* mov_imm to modify */
+ADDRTAKEN_LABEL(sandbox_immediate_addr_plus_four:)
+        mov      REG_XAX, HEX(1)
+        mov      REG_XBX, HEX(2)
+        lea      REG_XDI, SYMREF(usebx_immediate_addr_plus_four - 4)
+        stosd    /* stos selfmod write in another page */
+        mov      REG_XAX, REG_XBX
+        ret
+END_FUNC(FUNCNAME)
+#undef FUNCNAME
+
+ALIGN_WITH_NOPS(4096)
```

Comment:
So this assumes that a page is 4096.  PAGE_SIZE won't work here, so please add an assert in main() that PAGE_SIZE is <=4096: better that to fire than the test fails mysteriously and someone has to spend time figuring out why.

Could use 64K here (as a named constant) which covers 16K and 64K page sizes (mostly seen on ARM).